handler.py

```diff
@@ -14,38 +16,120 @@
+            "statusCode": 200,
+            "body": json.dumps({'OK': 'target added to the queue'})
+        }).with_security_headers()
+
+    else:
+        logger.error("Target Validation Failed of: " +
+                     json.dumps(target.targetname))
+        return Response({
+             "statusCode": 400,
+             "body": json.dumps({'error': 'target was not valid or missing'})
+         }).with_security_headers()
 
+
+def runScheduledPortScan(event, context):
     # For demo purposes, we will use a static list here
     target_list = [
         "www.mozilla.org",
         "infosec.mozilla.org",
     ]
     randomizer = Randomizer(target_list)
     target = Target(randomizer.next())
+
     if target.isValid():
-        logger.info("Tasking Observatory Scan of: " +
+        logger.info("Tasking Port Scan of: " +
                     json.dumps(target.targetname))
-        scanner = HTTPObservatoryScanner()
-        scan_result = scanner.scan(target)
-        logger.info("Completed Observatory Scan of " +
+        scanner = PortScanner()
+        results = scanner.scanTCP(target.targetname)
+        logger.info("Completed Port Scan of " +
                     json.dumps(target.targetname))
-        send_to_s3(target.targetname, scan_result)
+        send_to_s3(target.targetname + "_tcpscan", results)
     else:
         logger.error("Target Validation Failed of: " +
                      json.dumps(target.targetname))
 
 
-def runObservatoryScanFromQ(event, context):
+def addObservatoryScanToQueue(event, context):
+    data = json.loads(event['body'])
+    target = Target(data.get('target'))
+
+    if target.isValid():
+        print(SQS_CLIENT.send_message(
+              QueueUrl=os.getenv('SQS_URL'),
+              MessageBody="observatory|" + target.targetname
+              ))
+        return Response({
+            "statusCode": 200,
+            "body": json.dumps({'OK': 'target added to the queue'})
+        }).with_security_headers()
+    
+    else:
+        logger.error("Target Validation Failed of: " +
+                     json.dumps(target.targetname))
+        return Response({
+             "statusCode": 400,
+             "body": json.dumps({'error': 'target was not valid or missing'})
+         }).with_security_headers()
 
-    # This is how we process the hostname
-    # Demonstration purposes, in the final example 
-    # we should extract the hostname and perform a scan
 
+def runScanFromQ(event, context):
+    
+    # Read the queue
     for record, keys in event.items():
         for item in keys:
             if "body" in item:
-                logger.info(item['body'])
+                message = item['body']
+                scan_type, target = message.split('|')
+                if scan_type == "observatory":
+                    scanner = HTTPObservatoryScanner()
+                    scan_result = scanner.scan(target)
+                    send_to_s3(target + "_observatory", scan_result)
```

Comment:
There's something about concating these names and types together that makes me thing path traversal bug, but I can't put my finger on how exactly one would do it  assuming the hostname would have to validate first.

Reply:
Fair hunch to have, however the validation of hostname is currently saving us. I'd love to be proven wrong though, in which case I think we would consider hardening this even more.